src/journal/journald-server.c

```diff
@@ -1177,91 +1279,51 @@ int server_flush_to_var(Server *s, bool require_flag_file) {
         sd_journal_set_data_threshold(j, 0);
 
-        SD_JOURNAL_FOREACH(j) {
-                Object *o = NULL;
-                JournalFile *f;
-
-                f = j->current_file;
-                assert(f && f->current_offset > 0);
-
-                n++;
-
-                r = journal_file_move_to_object(f, OBJECT_ENTRY, f->current_offset, &o);
-                if (r < 0) {
-                        log_error_errno(r, "Can't read entry: %m");
-                        goto finish;
-                }
-
-                r = journal_file_copy_entry(f, s->system_journal->file, o, f->current_offset);
-                if (r >= 0)
-                        continue;
+        /* Try switching the system journal from volatile to persistent */
+        (void) system_journal_open(s, true, false);
 
-                if (!shall_try_append_again(s->system_journal->file, r)) {
-                        log_error_errno(r, "Can't write entry: %m");
-                        goto finish;
-                }
+        if (!s->system_journal)
+                return 0;       /* persistent mode disabled */
 
-                log_info("Rotating system journal.");
+        /* And switch user journals to persistent mode too */
+        old_user_journals = TAKE_PTR(s->user_journals);
+        s->user_journals = ordered_hashmap_new(NULL);
+        if (!s->user_journals)
+                return log_oom();
 
-                server_rotate(s);
-                server_vacuum(s, false);
+        log_debug("Flushing to %s...", s->system_storage.path);
 
-                if (!s->system_journal) {
-                        log_notice("Didn't flush runtime journal since rotation of system journal wasn't successful.");
-                        r = -EIO;
-                        goto finish;
-                }
+        r = do_flush(s, j);
+        if (r < 0) {
+                /* Something went wrong, let's switch back to volatile mode. */
+                ordered_hashmap_clear_with_destructor(s->user_journals, managed_journal_file_close);
+                s->user_journals = TAKE_PTR(old_user_journals);
+                s->system_journal = managed_journal_file_close(s->system_journal);
 
-                log_debug("Retrying write.");
-                r = journal_file_copy_entry(f, s->system_journal->file, o, f->current_offset);
-                if (r < 0) {
-                        log_error_errno(r, "Can't write entry: %m");
-                        goto finish;
-                }
+                return log_error_errno(r, "Flushing to %s failed: %m", s->system_storage.path);
         }
 
-        r = 0;
-
-finish:
-        if (s->system_journal)
-                journal_file_post_change(s->system_journal->file);
+        assert(s->system_journal);
+        journal_file_post_change(s->system_journal->file);
 
         s->runtime_journal = managed_journal_file_close(s->runtime_journal);
+        ordered_hashmap_clear_with_destructor(old_user_journals, managed_journal_file_close);
```

Comment:
for new code please don#t use the _with_destructor() versions of the hashmap and set destructors anymore. Instead provide a hash_ops structure with a constructure configured in there. That allows you to destroy the hashmap without specifying any manual destructor, and you can use `_cleanup_` and stuff